### PR TITLE
Proxy

### PR DESCRIPTION
**Pull request by zcstarr**
A lot of little features along the way here. Here's a quick overview of the changes to make proxy servicing work.

The proxy service is organized as follows 
**ProxyServer** -- Takes a Router and set of external facing ports for different protocols
**Router** - Takes a service notification queue 
**ServiceRunnerServer** - normal service runner things

Internally the proxy server 
works similar to GCP load balancing,
**frontends** -  the servers that will receive all requests
**backends** - the internal clients that forward or receive requests from services
**services** - the normal services that service runner manages    

**request path** 
client_ => _frontend_=>_backend_=> _service_ 
**response_path**
_service_ => _backend_=> _frontend_ => _client_

In between this is a connection manager that handles connection state transitions and enforcing any connection rules. 

**Additionally, the following features had to be added** 
1. integration api test end to end / developer friendly local file support for configuration extensions
2. exponential back off for starting services, that tests if a port is able to be bound. // This was added to remove confusion between startService starting and receiving an immediate connection failure, because the port was not yet listening. So now startService means you can make request to the underlying service without failing.
3. External service notifications events system, which allows events from the process manager to be subscribed to without fear of non critical systems messing with the events.
4. Programatic interface for service runner api - good for testing and for future menubar project to embed within the process.

Let me know if you have any questions/ what ya think? This was alot more involved, than anticipated, and the PR kind of sprawled a bit. I will probably squash this down to one or 2 commits before submitting. Thought it was more informative if I removed alot of the imm. commits.

### CODE REVIEW

**Comment by zcstarr** [comment]
https://jenkins.etclabs.org/blue/organizations/jenkins/jade-service-runner/detail/proxy/11/pipeline

The issue seems to be strange with regards to windows machine, maybe a github permissions issue?

File: src/lib/jsonRpcError.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+export const error = (message: string, id: number, data: any): JSONRpcError => {
+  let msg = message;
+  let code = errorMap.get(message);
+  if (code === undefined) {
+    code = -32000;
+    msg = "Unknown Error";
+  }
+  return { id, jsonrpc: "2.0", error: { code, message: msg, data },
+  };
```
Comment (shanejonas) [inline src/lib/jsonRpcError.ts]: formatting

File: src/lib/router.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import { makeLogger } from "./logging";
+interface RoutingInfo {
+  host: "localhost";
+  port: number;
+  protocol: util.Protocol;
+}
+const logger = makeLogger("ServiceRunner", "Router");
+/**
+ * Router - is the class that routes services via /serviceName/enviornment/version pattern
+ * the pattern can be made to be dynamic ini future by extending the resolution functoin to support
```
Comment (shanejonas) [inline src/lib/router.ts]:
```suggestion
 * the pattern can be made to be dynamic in the future by extending the resolution function to support
```

**Pull request merged**